### ISSUE TITLE
The click event request should be in async and not sync, on large servers it can be a problem

### ISSUE THREAD

**Issue by thiagogebrimm** [EvModder/DropHeads#72]
```
[00:41:12] [Watchdog Thread/ERROR]: --- DO NOT REPORT THIS TO PURPUR - THIS IS NOT A BUG OR A CRASH  - git-Purpur-1756 (MC: 1.19.2) ---
[00:41:12] [Watchdog Thread/ERROR]: The server has not responded for 10 seconds! Creating thread dump
[00:41:12] [Watchdog Thread/ERROR]: ------------------------------
[00:41:12] [Watchdog Thread/ERROR]: Server thread dump (Look for plugins here before reporting to Purpur!):
[00:41:12] [Watchdog Thread/ERROR]: Ticking entity: minecraft:player, entity class: net.minecraft.server.level.EntityPlayer
[00:41:12] [Watchdog Thread/ERROR]: Entity status: removed: false, valid: true, alive: true, is passenger: false
[00:41:12] [Watchdog Thread/ERROR]: Entity UUID: 20f487a9-a595-42e8-b892-a8197564b5ba
[00:41:12] [Watchdog Thread/ERROR]: Position: world: 'world' at location (127.72902768285753, 78.0, -3392.0268197989976)
[00:41:12] [Watchdog Thread/ERROR]: Velocity: (0.0, -0.0784000015258789, 0.0) (in blocks per tick)
[00:41:12] [Watchdog Thread/ERROR]: Entity AABB: AABB[127.4290276709366, 78.0, -3392.3268198109186] -> [128.02902769477845, 79.79999995231628, -3391.7268197870767]
[00:41:12] [Watchdog Thread/ERROR]: Total packets processed on the main thread for all players: 69145326
[00:41:12] [Watchdog Thread/ERROR]: ------------------------------
[00:41:12] [Watchdog Thread/ERROR]: Current Thread: Server thread
[00:41:12] [Watchdog Thread/ERROR]: 	PID: 43 | Suspended: false | Native: true | State: RUNNABLE
[00:41:12] [Watchdog Thread/ERROR]: 	Stack:
[00:41:12] [Watchdog Thread/ERROR]: 		java.base@17.0.3/sun.nio.ch.SocketDispatcher.read0(Native Method)
[00:41:12] [Watchdog Thread/ERROR]: 		java.base@17.0.3/sun.nio.ch.SocketDispatcher.read(SocketDispatcher.java:47)
[00:41:12] [Watchdog Thread/ERROR]: 		java.base@17.0.3/sun.nio.ch.NioSocketImpl.tryRead(NioSocketImpl.java:261)
[00:41:12] [Watchdog Thread/ERROR]: 		java.base@17.0.3/sun.nio.ch.NioSocketImpl.implRead(NioSocketImpl.java:312)
[00:41:12] [Watchdog Thread/ERROR]: 		java.base@17.0.3/sun.nio.ch.NioSocketImpl.read(NioSocketImpl.java:350)
[00:41:12] [Watchdog Thread/ERROR]: 		java.base@17.0.3/sun.nio.ch.NioSocketImpl$1.read(NioSocketImpl.java:803)
[00:41:12] [Watchdog Thread/ERROR]: 		java.base@17.0.3/java.net.Socket$SocketInputStream.read(Socket.java:976)
[00:41:12] [Watchdog Thread/ERROR]: 		java.base@17.0.3/sun.security.ssl.SSLSocketInputRecord.read(SSLSocketInputRecord.java:478)
[00:41:12] [Watchdog Thread/ERROR]: 		java.base@17.0.3/sun.security.ssl.SSLSocketInputRecord.readFully(SSLSocketInputRecord.java:461)
[00:41:12] [Watchdog Thread/ERROR]: 		java.base@17.0.3/sun.security.ssl.SSLSocketInputRecord.decodeInputRecord(SSLSocketInputRecord.java:243)
[00:41:12] [Watchdog Thread/ERROR]: 		java.base@17.0.3/sun.security.ssl.SSLSocketInputRecord.decode(SSLSocketInputRecord.java:181)
[00:41:12] [Watchdog Thread/ERROR]: 		java.base@17.0.3/sun.security.ssl.SSLTransport.decode(SSLTransport.java:111)
[00:41:12] [Watchdog Thread/ERROR]: 		java.base@17.0.3/sun.security.ssl.SSLSocketImpl.decode(SSLSocketImpl.java:1505)
[00:41:12] [Watchdog Thread/ERROR]: 		java.base@17.0.3/sun.security.ssl.SSLSocketImpl.readHandshakeRecord(SSLSocketImpl.java:1420)
[00:41:12] [Watchdog Thread/ERROR]: 		java.base@17.0.3/sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:455)
[00:41:12] [Watchdog Thread/ERROR]: 		java.base@17.0.3/sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:426)
[00:41:12] [Watchdog Thread/ERROR]: 		java.base@17.0.3/sun.net.www.protocol.https.HttpsClient.afterConnect(HttpsClient.java:580)
[00:41:12] [Watchdog Thread/ERROR]: 		java.base@17.0.3/sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.connect(AbstractDelegateHttpsURLConnection.java:183)
[00:41:12] [Watchdog Thread/ERROR]: 		java.base@17.0.3/sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1668)
[00:41:12] [Watchdog Thread/ERROR]: 		java.base@17.0.3/sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1592)
[00:41:12] [Watchdog Thread/ERROR]: 		java.base@17.0.3/sun.net.www.protocol.https.HttpsURLConnectionImpl.getInputStream(HttpsURLConnectionImpl.java:224)
[00:41:12] [Watchdog Thread/ERROR]: 		DropHeads.jar//net.evmodder.EvLib.extras.WebUtils.getReadURL(WebUtils.java:47)
[00:41:12] [Watchdog Thread/ERROR]: 		DropHeads.jar//net.evmodder.EvLib.extras.WebUtils.getGameProfile(WebUtils.java:211)
[00:41:12] [Watchdog Thread/ERROR]: 		DropHeads.jar//net.evmodder.DropHeads.listeners.BlockClickListener.getHeadNameData(BlockClickListener.java:128)
[00:41:12] [Watchdog Thread/ERROR]: 		DropHeads.jar//net.evmodder.DropHeads.listeners.BlockClickListener.getHeadNameData(BlockClickListener.java:148)
[00:41:12] [Watchdog Thread/ERROR]: 		DropHeads.jar//net.evmodder.DropHeads.listeners.BlockClickListener.onBlockClickEvent(BlockClickListener.java:227)
[00:41:12] [Watchdog Thread/ERROR]: 		com.destroystokyo.paper.event.executor.asm.generated.GeneratedEventExecutor633.execute(Unknown Source)
[00:41:12] [Watchdog Thread/ERROR]: 		org.bukkit.plugin.EventExecutor.lambda$create$1(EventExecutor.java:75)
[00:41:12] [Watchdog Thread/ERROR]: 		org.bukkit.plugin.EventExecutor$$Lambda$6806/0x000000080217c5b8.execute(Unknown Source)
[00:41:12] [Watchdog Thread/ERROR]: 		co.aikar.timings.TimedEventExecutor.execute(TimedEventExecutor.java:76)
[00:41:12] [Watchdog Thread/ERROR]: 		org.bukkit.plugin.RegisteredListener.callEvent(RegisteredListener.java:70)
[00:41:12] [Watchdog Thread/ERROR]: 		org.bukkit.plugin.SimplePluginManager.callEvent(SimplePluginManager.java:678)
[00:41:12] [Watchdog Thread/ERROR]: 		org.bukkit.craftbukkit.v1_19_R1.event.CraftEventFactory.callPlayerInteractEvent(CraftEventFactory.java:544)
[00:41:12] [Watchdog Thread/ERROR]: 		net.minecraft.server.level.ServerPlayerGameMode.useItemOn(ServerPlayerGameMode.java:527)
[00:41:12] [Watchdog Thread/ERROR]: 		net.minecraft.server.network.ServerGamePacketListenerImpl.handleUseItemOn(ServerGamePacketListenerImpl.java:2053)
[00:41:12] [Watchdog Thread/ERROR]: 		net.minecraft.network.protocol.game.ServerboundUseItemOnPacket.handle(ServerboundUseItemOnPacket.java:37)
[00:41:12] [Watchdog Thread/ERROR]: 		net.minecraft.network.protocol.game.ServerboundUseItemOnPacket.a(ServerboundUseItemOnPacket.java:9)
[00:41:12] [Watchdog Thread/ERROR]: 		net.minecraft.network.protocol.PacketUtils.lambda$ensureRunningOnSameThread$0(PacketUtils.java:51)
[00:41:12] [Watchdog Thread/ERROR]: 		net.minecraft.network.protocol.PlayerConnectionUtils$$Lambda$12270/0x00000008039a0930.run(Unknown Source)
[00:41:12] [Watchdog Thread/ERROR]: 		net.minecraft.server.TickTask.run(TickTask.java:18)
[00:41:12] [Watchdog Thread/ERROR]: 		net.minecraft.util.thread.BlockableEventLoop.doRunTask(BlockableEventLoop.java:153)
[00:41:12] [Watchdog Thread/ERROR]: 		net.minecraft.util.thread.ReentrantBlockableEventLoop.doRunTask(ReentrantBlockableEventLoop.java:24)
[00:41:12] [Watchdog Thread/ERROR]: 		net.minecraft.server.MinecraftServer.doRunTask(MinecraftServer.java:1387)
[00:41:12] [Watchdog Thread/ERROR]: 		net.minecraft.server.MinecraftServer.d(MinecraftServer.java:185)
[00:41:12] [Watchdog Thread/ERROR]: 		net.minecraft.util.thread.BlockableEventLoop.pollTask(BlockableEventLoop.java:126)
[00:41:12] [Watchdog Thread/ERROR]: 		net.minecraft.server.MinecraftServer.pollTaskInternal(MinecraftServer.java:1364)
[00:41:12] [Watchdog Thread/ERROR]: 		net.minecraft.server.MinecraftServer.pollTask(MinecraftServer.java:1357)
[00:41:12] [Watchdog Thread/ERROR]: 		net.minecraft.util.thread.BlockableEventLoop.runAllTasks(BlockableEventLoop.java:114)
[00:41:12] [Watchdog Thread/ERROR]: 		net.minecraft.server.MinecraftServer.tickServer(MinecraftServer.java:1491)
[00:41:12] [Watchdog Thread/ERROR]: 		net.minecraft.server.MinecraftServer.runServer(MinecraftServer.java:1213)
[00:41:12] [Watchdog Thread/ERROR]: 		net.minecraft.server.MinecraftServer.lambda$spin$1(MinecraftServer.java:310)
[00:41:12] [Watchdog Thread/ERROR]: 		net.minecraft.server.MinecraftServer$$Lambda$4425/0x000000080154f008.run(Unknown Source)
[00:41:12] [Watchdog Thread/ERROR]: 		java.base@17.0.3/java.lang.Thread.run(Thread.java:833)
[00:41:12] [Watchdog Thread/ERROR]: ------------------------------
```

**Comment by EvModder** [comment]
Good point, I'll see if I can do that.
In the meantime if it's causing issues, you can triage it by disabling or rate-limiting how often the plugin will process click events:

head-click-listener: true
head-click-max-rate-in-ticks: 10
(in config.yml)

**Comment by EvModder** [comment]
One downside to doing it asynchronously, is when the player lacks permission to view head info, DH currently does not cancel the click event, but this requires synchronously knowing the type of head being clicked (e.g., if they have permission to view info for player/mob heads but not custom-texture decoration heads).

This would be a relatively minor change in behavior, but could foreseeably be annoying to builders, etc.

Another option that would be relatively easy to implement would be a config setting to disable web-based UUID lookups for click events (the part of the code which checks if the account exists on Mojang's servers). This would only minorly affect things; specifically when clicking heads of inactive players whose server-side data is nonexistent or outdated (name/skin change since last login, etc).

Basically, DH would not be able to distinguish between "<Player who has never joined>'s Head" and "Custom Decoration Skull with no associated player account" but aside from that it would behave identically.

LMK what you think, if that is a good compromise for performance or not.
(side note, uuid-profile lookups are cached, so only initial click events for a skull should have potential to cause lag)

**Comment by thiagogebrimm** [comment]
> Another option that would be relatively easy to implement would be a config setting to disable web-based UUID lookups for click events (the part of the code which checks if the account exists on Mojang's servers). This would only minorly affect things; specifically when clicking heads of inactive players whose server-side data is nonexistent or outdated (name/skin change since last login, etc).

this option seems to be the most suitable

**Comment by EvModder** [comment]
Added this option in v3.7.7, which I will be uploading tonight or tomorrow

**Comment by EvModder** [comment]
Note I also did some major improvements to the profile lookup code (and caching), as well as adding the option above, so I think it's safe to mark this FR as complete.
